### ISSUE TITLE
Verilog grammar status

### ISSUE THREAD

**Issue by quantrpeter** [antlr/grammars-v4#3028]
Hi Mr @msagca
     How is the status of verilog grammar, is it stable now?
thanks
Peter

**Comment by msagca** [comment]
Hi @quantrpeter 

I think the grammar is in a good state. Of course it can be further optimized and improved but it works fine enough. As a user of it what do you think?

**Comment by quantrpeter** [comment]
I was using the old version to build our tool, I remember last year there is a big change, grammar changed a lot, and listener stub are all changed. Now I want to use it to build a logic synthesizer, that's why I ask if it is stable. thanks for creating it Mr Msagca.

**Comment by quantrpeter** [comment]
I just download a very big verilog, over 10000 line and test your grammar, it successful parsed it. thanks for your great effort, i can start building the logic synthesizer now.

**Comment by quantrpeter** [comment]
when do we need to use VerilogPreParser?

**Comment by msagca** [comment]
> when do we need to use VerilogPreParser?

It parses preprocessor directives. You need to handle them first before beginning the actual parse. If you're sure that the input text does not contain any directives that could impact the resulting parse tree then you can skip this step.

**Comment by quantrpeter** [comment]
thanks, if i put any text to parse for VerilogPreParser, it doesn't give me any error, may I know why please?

```
		VerilogLexer lexer = new VerilogLexer(CharStreams.fromString("hello this is fake"));
		CommonTokenStream tokenStream = new CommonTokenStream(lexer);
		VerilogPreParser parser = new VerilogPreParser(tokenStream);
		parser.source_text();
```

**Comment by msagca** [comment]
@quantrpeter

Tokens belonging to preprocessor directives are on the [DIRECTIVES](https://github.com/antlr/grammars-v4/blob/master/verilog/verilog/VerilogLexer.g4#L26) channel. You have to pass the correct channel ID (3) to [CommonTokenStream](https://www.antlr.org/api/Java/org/antlr/v4/runtime/CommonTokenStream.html).

**Comment by quantrpeter** [comment]
```
String str = IOUtils.toString(TestCompile.class.getClassLoader().getResourceAsStream("verilogexample/tb.v"), "utf-8");
VerilogLexer lexer = new VerilogLexer(CharStreams.fromString(str));
CommonTokenStream tokenStream = new CommonTokenStream(lexer, VerilogLexer.DIRECTIVES);
tokenStream.fill();
VerilogPreParser parser = new VerilogPreParser(tokenStream);
parser.source_text();
```

try to parse
```
module tb;
  reg clk, d, rstn;
  wire q;
  reg [3:0] delay;

  my_design u0 ( .clk(clk), .d(d),
`ifdef INCLUDE_RSTN asdasd asda s da
                .rstn(rstn),
`endif
                .q(q));

  always #10 clk = ~clk;

  initial begin

    {d, rstn, clk} = 0;

	#20 rstn = 1;
	
    for (integer i = 0 ; i < 20; i=i+1) begin
      delay = $random;
      #(delay) d = $random;
    end

    #20 $finish;
  end

endmodule
```

line 7 is fake, but no error message, any hints?

**Comment by msagca** [comment]
@quantrpeter 

IEEE 1364-2005 spec does not explicitly state that `ifdef_group_of_lines` should begin on a new line; so, I didn't require it in the grammar. Anything that comes after `INCLUDE_RSTN` (`text_macro_identifier`) is a part of `ifdef_group_of_lines`.

```ebnf
ifdef_directive ::=
`ifdef text_macro_identifier
ifdef_group_of_lines
{ `elsif text_macro_identifier elsif_group_of_lines }
[ `else else_group_of_lines ]
`endif
```

**Comment by quantrpeter** [comment]
nono, i typed fake text (after "INCLUDE_RSTN") in the ifdef, but it doesn't give me any error

```
`ifdef INCLUDE_RSTN asdasd asda s da
                .rstn(rstn),
`endif
```

**Comment by msagca** [comment]
@quantrpeter 

`VerilogParser` would detect that error. You've only parsed the compiler directives.

**Comment by msagca** [comment]
@quantrpeter 

If you mean this part: `asdasd asda s da`, then I explained why. Yes, it's not valid Verilog but we cannot detect this until we preprocess the input and parse it with `VerilogParser`.

**Comment by quantrpeter** [comment]
thanks @msagca , i understand now. But should we use VL instead of '|' in VerilogParser.g4? If not, we can't use ctx.expression().VL() to decide we have a '|' or not.

<img width="1441" alt="Screenshot 2023-02-13 at 1 57 06 AM" src="https://user-images.githubusercontent.com/50471577/218328366-353c05d5-2a1e-48b2-a30c-2af146d3b984.png">

**Comment by msagca** [comment]
@quantrpeter 

They correspond to the same token. If you open the generated `VerilogParser.tokens` or `VerilogLexer.tokens` file you will see that both `VL` and `|` have the same token ID (178).

**Comment by klcheungaj** [comment]
hi @msagca 
I also have a question about SystemVerilog and Verilog's preprocessor grammar.
Currently MACRO_USAGE and MACRO_NAME are defined as follows:
```
MACRO_NAME : IDENTIFIER ( [ \t\r\n]* MACRO_ARGS )? -> channel(DIRECTIVES), mode(MACRO_TEXT_MODE) ;
MACRO_USAGE : IDENTIFIER ( [ \t\r\n]* MACRO_ARGS )? -> channel(DIRECTIVES), popMode ;
```
but do we actually need \r and \n?
If the source code before preprocessing is like this:
```

`define MACRO1   
(a);
```
`(a)` will be regarded as part of the MACRO_NAME, while `;` becomes MACRO_TEXT. Both verilator and iverilog have following result:
```


(a);
```

Apart from that, I think it will be more convenient for users if macro identifier and macro arguments can be matched by different tokens rules so that we can have a parser rule to get the macro argument list.

**Comment by msagca** [comment]
@klcheungaj 

You're totally right about the newline character. It terminates the macro and shouldn't be allowed there. I'll fix this.

Separating macro identifier from macro arguments is tricky. I couldn't find a nice way of doing it while keeping the grammar target agnostic. I'll give it a try again. Maybe you can suggest an alternative solution.

**Comment by klcheungaj** [comment]
You are right. I didn't notice them. 

Sorry that I'm not good at parser. I can't think of a good way to implement it.
